### PR TITLE
Fix and prevent compatibility settings from becoming not effective 

### PR DESCRIPTION
**Pull request by cw5121**
When I set `compatibility = '20.8'`, settings `formatdatetime_f_prints_single_zero` and `formatdatetime_parsedatetime_m_is_month_name` are not set to the previous values.

This resulted in a query error during the version upgrade process due to version discrepancies between the shards.

before fix:

```sql
SELECT
    name,
    value,
    changed,
    default
FROM system.settings
WHERE name = 'compatibility'
┌─name──────────┬─value─┬─changed─┬─default─┐
│ compatibility │ 20.8  │       1 │         │
└───────────────┴───────┴─────────┴─────────┘

SELECT
    name,
    value,
    changed,
    default
FROM system.settings
WHERE name IN ('formatdatetime_parsedatetime_m_is_month_name', 'formatdatetime_f_prints_single_zero')
┌─name─────────────────────────────────────────┬─value─┬─changed─┬─default─┐
│ formatdatetime_f_prints_single_zero          │ 0     │       0 │ 0       │
│ formatdatetime_parsedatetime_m_is_month_name │ 1     │       0 │ 1       │
└──────────────────────────────────────────────┴───────┴─────────┴─────────┘
```

after fix:

```sql
SELECT
    name,
    value,
    changed,
    default
FROM system.settings
WHERE name IN ('formatdatetime_parsedatetime_m_is_month_name', 'formatdatetime_f_prints_single_zero')
┌─name─────────────────────────────────────────┬─value─┬─changed─┬─default─┐
│ formatdatetime_f_prints_single_zero          │ 1     │       1 │ 0       │
│ formatdatetime_parsedatetime_m_is_month_name │ 0     │       1 │ 1       │
└──────────────────────────────────────────────┴───────┴─────────┴─────────┘
```

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed a bug that compatibility level '23.4' was not properly applied.

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 4a438db6a5019d37c9fe8011f6fb7d751865983a with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65737/4a438db6a5019d37c9fe8011f6fb7d751865983a/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Cloud fork sync (only for ClickHouse Inc. employees)</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

**Comment by alexey-milovidov** [comment]
Let's add automatic checks for similar errors.

**Comment by cw5121** [comment]
> Let's add automatic checks for similar errors.

The reason the parameters are not effective here is that the values for the same key are being overwritten. Switching `settings_changes_history` from `std::map` to `std::multimap` can prevent this issue.

**Comment by alexey-milovidov** [comment]
Having a global variable with a non-trivial constructor (with memory allocation) is a problem. We should change it to something like `std::initializer_list`. Then do lazy initialization (e.g. function-level static) and check for duplicate keys there.

**Comment by cw5121** [comment]
@rschu1ze hi, I found it didn't meet expectations after modified. Could you help me, thanks.

**Comment by rschu1ze** [comment]
I'll check.

**Comment by cw5121** [comment]
CI seems not update.

File: src/Storages/System/StorageSystemSettingsChanges.cpp [hunk 1]
```diff
@@ -26,6 +26,7 @@ ColumnsDescription StorageSystemSettingsChanges::getColumnsDescription()
 
 void StorageSystemSettingsChanges::fillData(MutableColumns & res_columns, ContextPtr, const ActionsDAG::Node *, std::vector<UInt8>) const
 {
+    const auto & settings_changes_history = getSettingsChangesHistory();
```
Comment (cw5121) [inline src/Storages/System/StorageSystemSettingsChanges.cpp:29]: I don't understand why using `auto settings_changes_history` causes the current version's compatible parameters to be set to `previous_value`.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**